### PR TITLE
Python: Propagate taint through parse_qs

### PR DESCRIPTION
**Pull request by RasmusWL**
Encountered this in Gruyere. It was not essential to anything, but we _should_ handle it none the less.

reference: https://docs.python.org/3.8/library/urllib.parse.html#urllib.parse.parse_qs

### CODE REVIEW

**Comment by BekaValentine** [comment]
This all looks good. I take it that we were unable to track taint through the definitions of any of these things, or if we could, it wasn't enough b/c of not being able to analyze certain imports, etc.?

**Comment by RasmusWL** [comment]
What an excellent question! I did not think about it, simply found that we didn't propagate taint properly, and made a fix.

So I looked over the `urllib.parse` code more in depth. Currently the taint stops at the call the [`_coerce_args`](https://github.com/python/cpython/blob/ee40e4b8563e6e1bc2bfb267da5ffc9a2293318d/Lib/urllib/parse.py#L729) -- to handle that case, we would need to track taint through star args (tracked in https://github.com/github/codeql-python-team/issues/49)

If that was solved the taint would stop because we currently don't propagate taint through [`str.split`](https://github.com/python/cpython/blob/ee40e4b8563e6e1bc2bfb267da5ffc9a2293318d/Lib/urllib/parse.py#L739) (fix in https://github.com/Semmle/ql/pull/3302).

Once these two problems have been fixed, it would be interesting to see if we can handle this case out-of-the-box without any special handling.

**Comment by RasmusWL** [comment]
No matter what, we should also handle `parse_qsl` :+1:

**Comment by RasmusWL** [comment]
After adding support for `parse_qsl` I thought that this might be enough to get the taint for `parse_qs` working, but alas; although we figure out that `pairs` is tainted, neither `name` nor `value` gets tainted :disappointed: [src](https://github.com/python/cpython/blob/ee40e4b8563e6e1bc2bfb267da5ffc9a2293318d/Lib/urllib/parse.py#L695) I'll try to investigate this a bit more in a different issue/PR.

In conclusion, for now, we need both of these special cases.

File: python/ql/src/semmle/python/security/strings/External.qll [hunk 1]
```diff
@@ -181,6 +185,58 @@ private predicate urlparse(ControlFlowNode fromnode, CallNode tonode) {
...
+            parse_qsl = Value::named("six.moves.urllib.parse.parse_qsl")
+            or
+            // Python 2
+            parse_qsl = Value::named("urlparse.parse_qsl")
+            or
+            // Python 2 deprecated version of `urlparse.parse_qsl`
+            parse_qsl = Value::named("cgi.parse_qsl")
+            or
+            // Python 3
+            parse_qsl = Value::named("urllib.parse.parse_qsl")
```
Comment (tausbn) [inline python/ql/src/semmle/python/security/strings/External.qll:229]: We should maybe factor out all of this "Python 2, Python 2 (deprecated), Python 3" logic in a shared predicate `urllib_method(string name)` to avoid copy-pasting it everywhere?

Comment (RasmusWL) [reply python/ql/src/semmle/python/security/strings/External.qll:229]: I see your point and I sort of agree. However, notice that the _Python 2 (deprecated)_ is only for `parse_qsl` and `parse_qs`.

We could have a `urllib_parse_method` helper predicate to encapsulate six/py2/py3. If we had 10 special cases I would probably be totally on board with that idea, but personally I don't feel a strong urge to do anything about it.

If you feel strongly about it, I'd be happy to do it. I'm probably a 2 out of 10 on [The Sliding Scale of Giving a Fuck](https://capwatkins.com/blog/the-sliding-scale-of-giving-a-fuck).

Comment (tausbn) [reply python/ql/src/semmle/python/security/strings/External.qll:229]: Nah, that's fine. I don't feel strongly about it.
